### PR TITLE
perf(dsv4/hc_pre): fuse into one syncall task; fold the gate phase into Phase D

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary

Lands `hc_pre` as ONE full-occupancy `pl.spmd(24)` launch (24 AIC + 48 AIV == every 910B core) fused by hard mix-`syncall` barriers, replacing the previous **multi-scope split version** (10 separate `pl.spmd` / `pl.at` dispatches). This matches the qwen3 `fa_fused` pattern — one task, one AICPU dispatch instead of per-scope dispatch overhead. Split-K atomic-add fills idle cubes / AIV lanes at small (decode) `T`.

The kernel runs in **3 phases / 2 barriers**. The former Phase C (a standalone `rsqrt + scale + sigmoid` gate pass, published by a 3rd barrier) is **folded into Phase D**: the C→D handoff was cross-core (the phases stripe work over cores differently), which is the only reason it needed a barrier — but each Phase-D task recomputes the ONE gate it consumes from the barrier-2-published `mixes_raw` / `sq_sum_acc` on its **own core**. `rsqrt + sigmoid` is nearly free on this latency-bound kernel, so deleting the barrier is a net win.

Precision-neutral (identical math, recomputed in place). `comb_logits` / `post_pad_store` survive only as same-core scratch (assemble → load-back, no barrier) because their downstream `pl.load → pl.store` needs an HC_PAD-wide, 32B-aligned tile; the `pre` gate is consumed in tensor-world so it needs no buffer.

## Perf — vs upstream `main` (split version)

Device L2 swimlane (a2a3), full-kernel wall span, 3 runs each:

| mode | upstream main (split, 10 scopes) | this PR (fused, 2 barriers) | Δ |
|------|----------------------------------|-----------------------------|---|
| decode (B=4, S=2) | 72.4 µs | **38.4 µs** | **−47%  (1.9×)** |
| prefill (B=1, S=128) | 85.7 µs | **64.7 µs** | **−25%  (1.3×)** |

The bulk is the fusion itself (one task, no per-scope AICPU dispatch). Dropping the 3rd barrier via the Phase C→D fold is a further **~−8% decode / ~−11% prefill** on its own — measured as an ablation, the fused kernel with vs without that barrier — corroborated by the scheduler trace (fewer AICPU loop-iterations blocked on barriers: `pop_miss` 299→197, `loop_iter` 104→81 on decode).

## Validation

- Real device (a2a3, d0): **decode** (B=4,S=2) and **prefill** (B=1,S=128) — `x_mixed` / `post` / `comb` all pass the golden check, on both the split and fused versions.
- `ruff` + `check-english-only` pre-commit hooks pass.
- Sim CI stays skipped for `hc_pre` (`a2a3sim`/`a5sim` do not model the hard `syncall`; the kernel carries the `# ci: no-dep-gen` marker).

### CODE REVIEW

File: models/deepseek/v4/hc_pre.py [hunk 1]
```diff
@@ -327,201 +140,240 @@ def _hc_pre_prefill(
...
+                        x0 = pl.cast(x_flat[t0:t0 + T_TILE, 0 * D + d0:0 * D + d0 + D_CHUNK], target_type=pl.FP32)
+                        x1 = pl.cast(x_flat[t0:t0 + T_TILE, 1 * D + d0:1 * D + d0 + D_CHUNK], target_type=pl.FP32)
+                        x2 = pl.cast(x_flat[t0:t0 + T_TILE, 2 * D + d0:2 * D + d0 + D_CHUNK], target_type=pl.FP32)
+                        x3 = pl.cast(x_flat[t0:t0 + T_TILE, 3 * D + d0:3 * D + d0 + D_CHUNK], target_type=pl.FP32)
+                        y0 = pl.row_expand_mul(x0, pre0)
+                        y1 = pl.row_expand_mul(x1, pre1)
+                        y2 = pl.row_expand_mul(x2, pre2)
+                        y3 = pl.row_expand_mul(x3, pre3)
+                        y_tile = pl.add(pl.add(y0, y1), pl.add(y2, y3))
+                        x_mixed = pl.assemble(x_mixed, pl.cast(y_tile, target_type=pl.BF16), [t0, d0])
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply models/deepseek/v4/hc_pre.py]: Fixed in 3b8efc2 — restored `mode="rint"`. Confirmed regression: upstream's x_mixed write used `mode="rint"` and every other v4 BF16 cast does too. Device re-validated (decode + prefill), still passes golden.

[the remark this answers is not in the document]
Comment (Hzfengsy) [reply models/deepseek/v4/hc_pre.py]: Fixed in 3b8efc2 (same as the gemini comment above) — restored `mode="rint"` to match the RNE policy in the precision guide and the prior x_mixed write paths.

File: models/deepseek/v4/hc_pre.py [hunk 2]
```diff
@@ -67,19 +76,23 @@
...
-T_MAX = max(DECODE_BATCH * DECODE_SEQ, PREFILL_BATCH * PREFILL_SEQ)
+MIX_PAD = 32  # mix_hc (24) padded to a 32-wide cube N / vector row
+HC_PAD = 8  # hc (4) padded for 32B-aligned vector ops
 
-# tiling
+# Full-occupancy launch: NUM_CORES persistent blocks == physical AIC count (910B = 24), so a
+# hard mix-syncall reaches every AIC + AIV. Must equal the physical count exactly -- pypto does
+# NOT cap the launch, and >24 makes the runtime multi-round the blocks so some physical core is
+# in a later round when the barrier fires -> AICore timeout 507018.
+NUM_CORES = 24
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply models/deepseek/v4/hc_pre.py:86]: Guarded in 3b8efc2 — `__main__` now rejects `-p a5`/`a5sim` with a clear message instead of hanging. `NUM_CORES=24` is the 910B physical AIC count; a backend-aware participant count + re-tuning/re-validation for Ascend950 is out of scope for this perf PR, so failing fast is the safe fix.

File: models/deepseek/v4/hc_pre.py [hunk 3]
```diff
@@ -327,201 +140,240 @@ def _hc_pre_prefill(
...
+                    post_base = hc_reshaped[0:1, HC_MULT:HC_MULT + HC_PAD]
+                    post_scaled = pl.mul(pl.row_expand_mul(mixes_raw[t0:t0 + T_TILE, HC_MULT:HC_MULT + HC_PAD], inv_col), scale1)
+                    post_logits = pl.add(post_scaled, pl.col_expand(post_scaled, post_base))
+                    post_pad = pl.mul(pl.recip(pl.add(pl.exp(pl.neg(post_logits)), 1.0)), 2.0)
+                    # stash HC_PAD-wide in same-core scratch, load back as an aligned [8,8] tile,
+                    # then pl.store narrows the valid [8,4] into the 4-wide post output.
+                    post_pad_store = pl.assemble(post_pad_store, post_pad, [t0, 0])
+                    post_tile = pl.load(post_pad_store, [t0, 0], [COMB_T_TILE, HC_PAD],
+                                        valid_shapes=[COMB_T_TILE, HC_MULT], target_memory=pl.MemorySpace.Vec)
+                    pl.store(post_tile, [t0, 0], post)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply models/deepseek/v4/hc_pre.py]: Tried this, but it does not compile. `post_pad` is a tensor-world value (`mixes_raw` slice → sigmoid), and `pl.store` requires a Vec tile: `tile.store requires first argument to be a TileType, but got TensorType`. `set_validshape` narrows the valid region but doesn't materialize a tile. A direct store would also need a `[COMB_T_TILE, HC_MULT]` (8×4) FP32 tile, whose 16-byte rows `pto.alloc_tile` rejects (needs 32B alignment) — which is exactly why the HC_PAD-wide same-core scratch exists. So `post_pad_store` stays (same-core scratch, no barrier; `post` is tiny, 8×4 FP32). I did adopt the `COMB_T_TILE` consistency nit and added a comment explaining the constraint. See 3b8efc2.

File: models/deepseek/v4/hc_pre.py [hunk 4]
```diff
@@ -327,201 +140,240 @@ def _hc_pre_prefill(
...
+
+    # Cross-barrier intermediates: allocated ONCE in GM/HBM, live across the phases.
+    x_fp32 = pl.create_tensor([t_linear, HC_DIM], dtype=pl.FP32)  # FP32 activations for the pure-AIC matmul + rms
+    mixes_raw = pl.create_tensor([t_linear, MIX_PAD], dtype=pl.FP32)
+    # comb_logits / post_pad_store are now SAME-CORE scratch: the sinkhorn / write_post task
+    # writes its token-tile's gate then reads it back on the SAME core (no barrier). They are
+    # kept (not deleted) because both feed a downstream pl.load->pl.store that needs an
+    # HC_PAD-wide (32B-aligned) tile -- a narrow [T_TILE, HC_MULT] FP32 tile is 16B rows and
+    # pto.alloc_tile rejects it. pre (mix_x) is consumed in tensor-world, so it needs no buffer.
+    post_pad_store = pl.create_tensor([t_linear, HC_PAD], dtype=pl.FP32)  # HC_PAD-wide same-core scratch; write_post narrows to post
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply models/deepseek/v4/hc_pre.py:158]: See the reply on the store site below — the `set_validshape` → direct `pl.store` doesn't compile (`pl.store` needs a Vec TileType; the gate is tensor-world), so `post_pad_store` is kept as same-core scratch. 3b8efc2.
